rule/rule.go

```diff
@@ -966,7 +984,19 @@ func (r *Rule) sync() {
 	}
 
 	call := r.expr.(*bzl.CallExpr)
-	call.X.(*bzl.Ident).Name = r.kind
+
+	var properName string = r.kind
```

Comment:
@fmeum could you please be more precise? 
Don't know how this would help in this particular case. `r.kind` has to be properly set to `(*bzl.Ident).Name` which is also the `string`. Removing all after `.` (dot) from `r.kind` is needed in case any will set the rule kind with `structName.macroName` rather than just `structName`. 
I think there is no reason now to limit rule kinds to be settable just to `structName` (without dot).

Reply:
Sorry, I was referring to making the `kind` field of `Rule` a `bzl.Expr` so that it can support arbitrarily nested dot expressions without having to go back and forth between `Expr` and string representations thereof.

I somehow wasn't able to open a PR on your branch, so this is the diff I came up with:

```diff
diff --git a/rule/rule.go b/rule/rule.go
index 8863a2e..5461430 100644
--- a/rule/rule.go
+++ b/rule/rule.go
@@ -713,7 +713,7 @@ func (l *Load) sync() {
 // Rule represents a rule statement within a build file.
 type Rule struct {
 	stmt
-	kind    string
+	kind    bzl.Expr
 	args    []bzl.Expr
 	attrs   map[string]*bzl.AssignExpr
 	private map[string]interface{}
@@ -721,10 +721,11 @@ type Rule struct {
 
 // NewRule creates a new, empty rule with the given kind and name.
 func NewRule(kind, name string) *Rule {
-	call := &bzl.CallExpr{X: &bzl.Ident{Name: kind}}
+	kindIdent := &bzl.Ident{Name: kind}
+	call := &bzl.CallExpr{X: kindIdent}
 	r := &Rule{
 		stmt:    stmt{expr: call},
-		kind:    kind,
+		kind:    kindIdent,
 		attrs:   map[string]*bzl.AssignExpr{},
 		private: map[string]interface{}{},
 	}
@@ -740,32 +741,26 @@ func NewRule(kind, name string) *Rule {
 	return r
 }
 
-func ruleFromExpr(index int, expr bzl.Expr) *Rule {
-	call, ok := expr.(*bzl.CallExpr)
-	if !ok {
-		return nil
+func isNestedDotOrIdent(expr bzl.Expr) bool {
+	if _, ok := expr.(*bzl.Ident); ok {
+		return true
 	}
-
-	var x *bzl.Ident
-	var suffix string = ""
-	var kind string
-
-	d, ok := call.X.(*bzl.DotExpr)
-	if ok {
-		x, ok = d.X.(*bzl.Ident)
-		suffix = d.Name
-	} else {
-		x, ok = call.X.(*bzl.Ident)
+	dot, ok := expr.(*bzl.DotExpr)
+	if !ok {
+		return false
 	}
+	return isNestedDotOrIdent(dot.X)
+}
 
+func ruleFromExpr(index int, expr bzl.Expr) *Rule {
+	call, ok := expr.(*bzl.CallExpr)
 	if !ok {
 		return nil
 	}
 
-	if len(suffix) > 0 {
-		kind = x.Name + "." + suffix
-	} else {
-		kind = x.Name
+	kind := call.X
+	if !isNestedDotOrIdent(kind) {
+		return nil
 	}
 
 	var args []bzl.Expr
@@ -800,12 +795,12 @@ func (r *Rule) ShouldKeep() bool {
 
 // Kind returns the kind of rule this is (for example, "go_library").
 func (r *Rule) Kind() string {
-	return r.kind
+	return bzl.FormatString(r.kind)
 }
 
 // SetKind changes the kind of rule this is.
 func (r *Rule) SetKind(kind string) {
-	r.kind = kind
+	r.kind = &bzl.Ident{Name: kind}
 	r.updated = true
 }
 
@@ -984,18 +979,7 @@ func (r *Rule) sync() {
 	}
 
 	call := r.expr.(*bzl.CallExpr)
-
-	var properName string = r.kind
-	if idx := strings.Index(properName, "."); idx != -1 {
-		properName = properName[:idx]
-	}
-
-	d, ok := call.X.(*bzl.DotExpr)
-	if ok {
-		d.X.(*bzl.Ident).Name = properName
-	} else {
-		call.X.(*bzl.Ident).Name = properName
-	}
+	call.X = r.kind
 
 	if len(r.attrs) > 1 {
 		call.ForceMultiLine = true


```

Reply:
Indeed this looks a bit better :smile: Thank you for the help and diff.